californium-core/src/main/java/org/eclipse/californium/core/network/CoapEndpoint.java

```diff
@@ -304,37 +315,66 @@ public CoapEndpoint(final InetSocketAddress address, final NetworkConfig config,
 	}
 
 	/**
-	 * Creates a new endpoint for a connector, configuration, message exchange and observation store.
+	 * Creates a new endpoint for a connector, configuration, message exchange
+	 * and observation store.
 	 * <p>
-	 * The endpoint will support the connector's implemented scheme and will bind to
-	 * the IP address and port the connector is configured for.
+	 * Intended to be called either by the {@link CoapEndpointBuilder} or a
+	 * subclass constructor. The endpoint will support the connector's
+	 * implemented scheme and will bind to the IP address and port the connector
+	 * is configured for.
 	 *
 	 * @param connector The connector to use.
+	 * @param applyConfiguration if {@code true}, apply network configuration to connector.
+	 *            Requires a {@link UDPConnector}.
 	 * @param config The configuration values to use.
-	 * @param store The store to use for keeping track of observations initiated by this
-	 *              endpoint.
-	 * @param exchangeStore The store to use for keeping track of message exchanges.
+	 * @param store The store to use for keeping track of observations initiated
+	 *            by this endpoint.
+	 * @param exchangeStore The store to use for keeping track of message
+	 *            exchanges.
 	 * @param endpointContextMatcher endpoint context matcher for relating
 	 *            responses to requests. If <code>null</code>, the result of
 	 *            {@link EndpointContextMatcherFactory#create(Connector, NetworkConfig)}
 	 *            is used as matcher.
+	 * @throws IllegalArgumentException if applyConfiguration is {@code true},
+	 *             but the connector is not a {@link UDPConnector}
 	 */
-	public CoapEndpoint(Connector connector, NetworkConfig config, ObservationStore store,
-			MessageExchangeStore exchangeStore, EndpointContextMatcher endpointContextMatcher) {
+	protected CoapEndpoint(Connector connector, boolean applyConfiguration, NetworkConfig config,
+			ObservationStore store, MessageExchangeStore exchangeStore, EndpointContextMatcher endpointContextMatcher) {
 		this.config = config;
 		this.connector = connector;
 		this.connector.setRawDataReceiver(new InboxImpl());
-		this.scheme =  CoAP.getSchemeForProtocol(connector.getProtocol());
-		MessageExchangeStore localExchangeStore = (null != exchangeStore) ? exchangeStore : new InMemoryMessageExchangeStore(config);
+		this.scheme = CoAP.getSchemeForProtocol(connector.getProtocol());
+
+		// when remove the deprecated constructors, this checks and defaults maybe removed also
+		MessageExchangeStore localExchangeStore = (null != exchangeStore) ? exchangeStore
+				: new InMemoryMessageExchangeStore(config);
 		ObservationStore observationStore = (null != store) ? store : new InMemoryObservationStore();
 		if (null == endpointContextMatcher) {
 			endpointContextMatcher = EndpointContextMatcherFactory.create(connector, config);
 		}
+
+		// keep for subclasses
+		if (applyConfiguration) {
```

Comment:
I would move this to `builder.build()` and so limit this `applyConfiguration` to the builder.

Reply:
The pain of such a move would be that the idea of sub-classing the CoapEndpoint would then not longer benefit from that feature. For the CoapTestEndpoint we could go for that, so, should we limit the sub-classing in the javadoc?

Reply:
A possibility : 
1. move this code in the builder.
2. add builder static utility  function which set UDPConnector from config and which is used by the builder.
3. use this utility function in CoapTestEndpoint.

The idea is to well separate responsibility. Now we have a builder, it is responsible to build endpoint. So Endpoint constructor should be simple. By the way I will remove null checking + default value from CoapEndpoint constructor.

This just a point of view, not a big deal if you don't share it and you prefer to keep the code like this ;)

Reply:
That solution looks like the previous one (current one, without this PR).

FMPOV, it seems to be more self-documented to have a "applyConfiguration" switch in the parameter list of the constructor than using a separate method.

The builder pattern and sub-classing doesn't really match, that's the main cause of this complexity.